idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/ui/dialogs/DialogSyntax.kt

```diff
@@ -0,0 +1,106 @@
+  fun IdeMetaPlugin.addFileAction(
+    actionId: String,
+    createText: String,
+    actionDescription: String,
+    buildDialog: CreateFileFromTemplateDialog.Builder.(project: Project, directory: PsiDirectory) -> Unit = Noop.effect3,
+    postProcess: (createdElement: PsiFile, templateName: String, customProperties: MutableMap<String, String>) -> Unit = Noop.effect3,
+    fileIcon: Icon? = KotlinFileType.INSTANCE.icon,
+    actionName: (directory: PsiDirectory, newName: String, templateName: String) -> String =
+      { _, _, _ -> createText },
+    createFileFromTemplate: (name: String, template: FileTemplate, dir: PsiDirectory) -> PsiFile? =
+      { name, template, dir -> NewKotlinFileAction.createFileFromTemplateWithStat(name, template, dir) },
+    startInWriteAction: Boolean = false
+  ): ExtensionPhase = // TODO: Investigate system when there is no `postProcess`
+    addAnAction(actionId, newFileAction(createText, actionDescription, buildDialog, postProcess, fileIcon, actionName, createFileFromTemplate, startInWriteAction))
+
+  fun DialogSyntax.newFileAction(
+    createText: String,
+    actionDescription: String,
+    buildDialog: CreateFileFromTemplateDialog.Builder.(project: Project, directory: PsiDirectory) -> Unit = Noop.effect3,
+    postProcess: (createdElement: PsiFile, templateName: String, customProperties: MutableMap<String, String>) -> Unit = Noop.effect3,
+    fileIcon: Icon? = KotlinFileType.INSTANCE.icon,
+    actionName: (directory: PsiDirectory, newName: String, templateName: String) -> String =
+      { _, _, _ -> createText },
+    createFileFromTemplate: (name: String, template: FileTemplate, dir: PsiDirectory) -> PsiFile? =
+      { name, template, dir -> NewKotlinFileAction.createFileFromTemplate(name, template, dir) },
+    startInWriteAction: Boolean = false
+  ): CreateFileFromTemplateAction =
+    object : CreateFileFromTemplateAction(createText, actionDescription, fileIcon), DumbAware {
+      override fun getActionName(directory: PsiDirectory?, newName: String, templateName: String?): String? =
+        directory?.let { d -> templateName?.let { t -> actionName(d, newName, t) } }
+
+      override fun buildDialog(project: Project?, directory: PsiDirectory?, builder: CreateFileFromTemplateDialog.Builder?): Unit =
+        project?.let { p -> directory?.let { d -> builder?.let { b -> buildDialog(b, p, d) } } } ?: Unit
+
+      override fun startInWriteAction(): Boolean = startInWriteAction
+
+      override fun isAvailable(dataContext: DataContext?): Boolean =
+        dataContext?.let { ctx ->
+          if (super.isAvailable(ctx))
+            LangDataKeys.IDE_VIEW.getData(ctx)?.let { ideView: IdeView ->
+              PlatformDataKeys.PROJECT.getData(ctx)?.let { project ->
+                ideView.directories.any {
+                  ProjectRootManager.getInstance(project).fileIndex.isInSourceContent(it.virtualFile)
+                }
+              }
+            }
+          else
+            false
```

Comment:
nit: if we have each case on the next line for both `if` and `else` perhaps we can go ahead and wrap them in brackets. It's more common in Kotlin to have the following conventions:

```
if (something) {
   something_happens
} else {
    something_else_happens
}
```

or 

```
if (something) something_happens else something_else_happens